### PR TITLE
Perform ssl-redirect after matching host

### LINKED ISSUE #691: Perform ssl-redirect after matching host
From Rafael Angeline in Slack:

I'm having problems with http.noTLS setting, since even when not setting a key at tls.hosts & putting http.noTLS: true at rules, those lines are generated at ConfigMap:
```frontend http-80
    bind *:80 
    mode http

    # Limit Connections
    acl is_proxy_https hdr(X-Forwarded-Proto) https
    redirect scheme https code 301 if ! is_proxy_https
    
    [...]
```
`redirect scheme https code 301 if ! is_proxy_https` this redirects my http requests with a 301 to https, which I do not think is the desired behavior - could you please advise?

Basically, I'm trying to serve this at my ingress:
```- host: donotforgettogohome.getschema.com
    http:
      noTLS: true
      paths:
      - backend:
          serviceName: donotforget-app
          servicePort: '80'
  - host: getschema.com
    http:
      paths:
        - path: '/'
          backend:
            serviceName: app-service
            servicePort: '80'
```

with a tls like:
```spec:
  tls:
  - hosts:
    - getschema.com
    secretName: app-service-tls
```

which means only getschema.com would be served with TLS and normal http to  donotforgettogohome.getschema.com ~ but that's not working since it always try to redirect to HTTPS.

editing manually the configmap and removing that line ensure things act normally. So will this be fixed by a next release? Also, do not forget one main question I have about setting the external IP of the service started by Ingress.

### PR DESCRIPTION
**Pull request by tamalsaha**
Fixes #691

### CODE REVIEW

**Comment by tamalsaha** [comment]
```
FP 80, mode HTTP

443( defined in YAML), mode HTTP
----------------------------------------------------
Redirect for Port 443 hosts  http->https for abc.com

```
tls:
 - hosts:
    - abc.com
rules:
 - host: abc.com --> 443
   paths: /about-us
 - host: xyz.com                                (Must NOT Redirect)
   paths: /about-us
```

Host: <domain>


GET HTTP/1 /about-us
Host: abc.com



Browser -------------------------------------> HAProxy (redirect)
        <----R 301 Location: https://abc.com----

Browser                                        (443)
301 (permanent) ------https://abc.com-------> HAProxy ---->pass-through--------------> backend pod
                ------------ENC--------->  TLS/Term -----------> HTTP(unenc)--------->
                -------ENC-----ELB --unenc--->  TLS/Term -----------> HTTP(unenc)---->


B http://  -(https://)-------------> HA




xyz.com (IP == abc.com)
Browser ----. http://xyz.com------> HAProxy -----------> backend pods
```

File: pkg/ingress/parser.go [hunk 1]
```diff
@@ -335,18 +343,55 @@ func (c *controller) generateConfig() error {
...
+				}
+			}
+
+			info := &httpInfo{Hosts: make(map[string][]*haproxy.HTTPPath)}
+			if v, ok := httpServices[binder]; ok {
+				info = v
+			} else {
+				httpServices[binder] = info
+			}
+			info.OffloadSSL = offloadSSL
```
Comment (diptadas) [inline pkg/ingress/parser.go]: Can't user define two different rules with same host, address and port but different paths?
In such case what if one rule has specified `NoTLS` but other rule don't?

Comment (tamalsaha) [reply pkg/ingress/parser.go]: User can define that but it is not supported by HAProxy. The offloadSSL-ing happens for `addr;port` bind line.

**Comment by tamalsaha** [comment]
Only 5 tests are failing now:

```
Summarizing 5 Failures:

[Fail] IngressTLS [JustBeforeEach] SSL Passthrough Should Open 443 with HTTP 
/home/tamal/go/src/github.com/appscode/voyager/test/e2e/ingress_tls.go:61

[Fail] IngressUpdates Secret Changed [It] Should update when secret changed 
/home/tamal/go/src/github.com/appscode/voyager/test/e2e/ingress_updates.go:109

[Fail] IngressOperations Create With custom target annotations [It] Should persist service IP 
/home/tamal/go/src/github.com/appscode/voyager/test/e2e/ingress_ops.go:207

[Fail] IngressOperations Update [It] Should add TCP rule 
/home/tamal/go/src/github.com/appscode/voyager/test/e2e/ingress_ops.go:511

[Fail] IngressOperations With Pod MaxConnections (1) Specified [It] Should Allow 1 Connection Concurrently 
/home/tamal/go/src/github.com/appscode/voyager/test/e2e/ingress_ops.go:796

Ran 77 of 84 Specs in 3310.690 seconds
FAIL! -- 72 Passed | 5 Failed | 0 Pending | 7 Skipped --- FAIL: TestE2E (3311.09s)
FAIL
```

**Pull request merged**